build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,289 @@
+
+/**
+ * Attempts to extract the type definition file path from the 'exports' field of a given package.json.
+ * Checks both 'import' and 'require' resolution methods to find the appropriate path.
+ * If the path is found, it is returned. Otherwise, an error is thrown.
+ * @param previousPackageJson - An object representing the previous package json
+ * @param previousBasePath - A string representing the path to the root of a package
+ * @returns A type definition filepath based on the appropriate export, or undefined if it cannot be found.
+ */
+export function getTypePathFromExport(
+	previousPackageJson: PackageJson,
+	previousBasePath: string,
+): string | undefined {
+	if (!previousPackageJson.exports) {
+		console.warn("The 'exports' field is missing in the package.json.");
+		return undefined;
+	}
+
+	const extractTypesPath = (exportEntry: any): string | undefined => {
+		return exportEntry?.types ? path.join(previousBasePath, exportEntry.types) : undefined;
+	};
+	const defaultSubpath = ".";
+	const exportEntry = previousPackageJson.exports[defaultSubpath];
+
+	// Check both 'import' and 'require' resolution methods
+	const typeDefinitionFilePath =
+		extractTypesPath(exportEntry?.import) ?? extractTypesPath(exportEntry?.require);
+	if (!typeDefinitionFilePath) {
+		// If no valid path is found, throw an error
+		throw new Error(
+			`Type definition file path could not be determined from the 'exports' field of '${getPreviousPackageJsonPath(
+				previousBasePath,
+			)}' using the default export entry '.'`,
+		);
+	}
+	return typeDefinitionFilePath;
+}
+/**
+ * Checks the package.json's exports entries and types field for a type definition filepath
+ * @returns string representing type definition file path
+ */
+export function getTypeDefinitionFilePath(packageBasePath: string): string | undefined {
+	const previousPackageJsonPath = getPreviousPackageJsonPath(packageBasePath);
+	const packageJson: PackageJson = readJsonSync(previousPackageJsonPath);
+	// Check the exports entries
+	if (packageJson.exports) {
+		return getTypePathFromExport(packageJson, packageBasePath);
+		// Check the types field from the previous package.json as a fallback
+	} else if (packageJson.types) {
+		return path.join(packageBasePath, packageJson.types);
+	} else {
+		throw new Error(
+			`Type definition file path could not be determined from '${previousPackageJsonPath},
+			)}'. No 'exports' nor 'type' fields found.`,
+		);
+	}
+}
+
+/**
+ * Extracts type data from a TS source file and creates a map where each key is a type name and the value is its type data.
+ * @param file - The source code file containing type data
+ * @returns The mapping between item and its type
+ */
+export function typeDataFromFile(file: SourceFile): Map<string, TypeData> {
+	const typeData = new Map<string, TypeData>();
+	const exportedDeclarations = file.getExportedDeclarations();
+
+	for (const declarations of exportedDeclarations.values()) {
+		for (const dec of declarations) {
```

Comment:
Nit (I generally recommend avoid abbreviations where possible, as an accessibly consideration):
```suggestion
		for (const declaration of declarations) {
```